File: src/main/java/de/omilke/bankingfx/report/savings/model/Category.java
package de.omilke.bankingfx.report.savings.model;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.SortedSet;
import java.util.TreeSet;

public class Category {

    private final String name;

    private BigDecimal sum = BigDecimal.ZERO;
    private final SortedSet<Entry> entries = new TreeSet<>(ENTRY_COMPARATOR);

    public Category(String name) {
        this.name = name;
    }

    public void addEntry(Entry entry) {

        entries.add(entry);
        this.sum = this.sum.add(entry.getAmount());
    }

    public String getName() {
        return name;
    }

    public SortedSet<Entry> getEntries() {
        return entries;
    }

    public BigDecimal getSum() {
        return sum;
    }

    /**
     * Provides the lowercase name for building a case-insensitive comparator.
     */
    public String getLowerCaseName() {
        return this.name.toLowerCase();
    }

    private static final Comparator<Entry> ENTRY_COMPARATOR = Comparator
            .comparing(Entry::getEntryDate).reversed()
            .thenComparing(Entry::getAmount)
            .thenComparing(Entry::getComment);

}

File: src/main/java/de/omilke/bankingfx/report/savings/model/Category.kt
package de.omilke.bankingfx.report.savings.model

import java.math.BigDecimal
import java.util.*

class Category(val name: String) : Comparable<Category> {

    private var _sum = BigDecimal.ZERO

    val sum: BigDecimal
        get() = _sum

    val entries: SortedSet<Entry> = TreeSet(SavingEntryComparator())

    fun addEntry(entry: Entry) {

        entries.add(entry)
        _sum = _sum.add(entry.getAmount())
    }

    /**
     * Provides the lowercase name for building a case-insensitive comparator.
     */
    private val lowerCaseName: String
        get() = name.toLowerCase()

    override fun compareTo(other: Category): Int {
        return compareValuesBy(
                this,
                other,
                compareBy(Category::sum).thenComparing(Category::lowerCaseName),
                { it })
    }
}